tests: smp: add some module and integration test cases

enjiamai opened the pull request:
Add some module and integration test cases for testing SMP:

There are several test cases added in this PR for testing:

Add module test case:
- test_smp_release_global_lock() and  test_smp_release_global_lock_irq()
   to verify z_smp_release_global_lock() works

Add integration test cases: 
- test_inc_concurrency()
   verify parallelly increase operations will fail if not applying synchronization on SMP.


Ps:  After clarified by Andy in issue #33551, I remove the test_smp_irq_lock_sem() and test_smp_irq_lock_mutex() because it's better to no use them in this way.

Signed-off-by: Enjia Mai <enjiax.mai@intel.com>

LeiW000 commented:
@Zhaoningx , please review the PR.

nashif reviewed (changes requested):
why are we testing those internal APIs directly? Are we trying to hit coverage? Could this be due to dead code? I would prefer to see the functionality and coverage being tested using the public APIs

enjiamai commented:
> why are we testing those internal APIs directly? Are we trying to hit coverage? Could this be due to dead code? I would prefer to see the functionality and coverage being tested using the public APIs

Got it, I understand now. I will correct this and not use this internal API for testing. Thank you!

LeiW000 commented on tests/kernel/smp/src/main.c:
```diff
@@ -624,6 +630,573 @@ void test_smp_ipi(void)
+ * z_smp_release_global_lock() to be called.
+ */
+void test_smp_release_global_lock_irq(void)
+{
```
The test case will always succeed. How do we know z_smp_global_lock() has been called.

Same to test_smp_release_global_lock().

enjiamai replied:
@LeiW000 Yes, you are right. There should be some checkpoint to see if the condition meets or not. Let me try to refine it.
Thanks for pointing out this!

Commit:
tests: smp: add some module and integration test cases

This PR add 2 module test cases:
- test_smp_release_global_lock() and test_smp_release_global_lock_irq()
  verify z_smp_release_global_lock() works.

And 1 integration test cases:
- test_inc_concurrency() to verify parallelly increase operations will
  fail if not applying synchronization on SMP.

Signed-off-by: Enjia Mai <enjiax.mai@intel.com>

enjiamai commented:
@nashif  @LeiW000 , I have addressed the comments, could you please review this PR again, thanks!

enjiamai commented:
Sorry for did not give it a clear description after the update. I have updated those test cases to better testing scenarios, and also not using the internal API anymore just for coverage. There is no dead code currently(the previous merged PR#33293 has delete the dead code of z_smp_reacquire_global_lock()).

So could you please help to review it again,  @nashif  @andyross, thank you!

andyross commented on tests/kernel/smp/src/main.c:
```diff
@@ -717,6 +739,487 @@ void test_workq_on_smp(void)
+ * @details Validate if the irq_lock() is called by a thread, then it
+ * allows another thread to run (such as giving a semaphore or sleeping
+ * for N milliseconds), the interrupt lock no longer applies and
+ * interrupts may be re-enabled while other processing occurs.
```
Yikes, no.  It's true that this has historically worked with irq_lock (because it was recursive, it wasn't possible for the kernel to know whether it was nested or not).  But that was effectively a bug.  Applications that get surprised when the kernel magically releases their locks tend to have terrible race conditions.

We do **not** want to test this condition, there is actually a github issue out there for me to make this an assertion so applications get a proper warning.

enjiamai replied:
Appreciate your review ^^, I got it and remove this incorrect test case.

andyross commented on tests/kernel/smp/src/main.c:
```diff
@@ -717,6 +739,487 @@ void test_workq_on_smp(void)
+ * @ingroup kernel_smp_tests
+ *
+ * @details Validate the scenario that make the internal APIs of SMP
+ * z_smp_release_global_lock() to be called.
```
Seems like there should be some way to exercise this with just plain irq_lock/unlock().  But whiteboxing this isn't awful, it's a complicated internal state and an exerciser with the public API would be no less confusing.

enjiamai replied:
Yes, got your point. I believe that there is some better way that uses less code but I spent lots of this on this without design the better testcase to cover it. I will keep finding them out. So can I just keep them at this time? Because they can cover all the branches of the function z_smp_release_global_lock(). Thank you!

andyross commented on tests/kernel/smp/src/main.c:
```diff
@@ -717,6 +739,487 @@ void test_workq_on_smp(void)
+ * assertion will be triggered. Because current context can be
+ * preempted and migrated to another SMP CPU during getting the
+ * current CPU. Using this is more safe than arch_curr_cpu() in
+ * context switch code.
```
We're writing a test to validate an assertion?  Isn't that circular?  Don't we now need a separate test case to validate the test that validates the assertion?  Only then can we get started on the test to validate the test case to validate the test that validates the assertion.

Just remove.  That assertion is there and works to protect against misuse.  And I know because I misuse the API all the time and trip over it.

enjiamai replied:
Got it, @andyross I removed this test case. But here I still try to cover the branch of _current_cpu. So I just try to put the validation of the assertion in another existing testcase test_get_cpu(), do you think is this acceptable?  Thank you!

andyross replied:
Again, this is being IMHO senselessly strict about code coverage.  An assertion is something that should "never happen".  So **by definition**, it represents uncovered code, and it always will.  All assertions work like that.  We'll spend forever chasing them down trying to "cover" them, and that's completely missing the point.

enjiamai replied:
Thanks for detailing this concept, so helpful. I think this corrects some of my misunderstandings. I just removed them completely.

andyross reviewed (changes requested):
Two cases to remove.  One that seems needless, the other is validating a behavior that is a historical bug that shouldn't even be valid.

enjiamai commented:
Hi, @andyross, Thanks so much for your review. I try to address the comments you gave, could you please take a look at it again?

andyross reviewed (approved):
Can't find anything more to complain about.  Thanks!

enjiamai commented:
Excuse me, @nashif , could you please take a look at this PR again? I have addressed all the comments and I think it can be merged, thank you!
